[DOCS] Clarify use of CCS on ML nodes

lcawl opened the pull request:
Related to https://github.com/elastic/elasticsearch/pull/66533

This PR updates https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-node.html to suggest adding the remote_cluster_client role to dedicated machine learning and transform nodes.

### Preview

* https://elasticsearch_66616.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/modules-node.html#ml-node
* https://elasticsearch_66616.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/ml-settings.html
* https://elasticsearch_66616.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/transform-settings.html

lcawl committed:
[DOCS] Clarify use of CCS on ML nodes

elasticmachine commented:
Pinging @elastic/ml-core (:ml)

droberts195 commented on docs/reference/modules/node.asciidoc:
```diff
@@ -340,6 +340,9 @@ xpack.ml.enabled: true <1>
 <1> The `xpack.ml.enabled` setting is enabled by default.
 
+If you want the node to use {ccs}, add `remote_cluster_client` to the list of
+roles. See <<remote-node>>.
```
I think we need some clarification on whether a node with `ml` and `remote_cluster_client` is still considered a dedicated ML node.  I think it should still be:

```suggestion
roles. See <<remote-node>>.  A node with both the `ml` and `remote_cluster_client` roles
but no others is still considered a dedicated {ml} node.
```

droberts195 commented on docs/reference/modules/node.asciidoc:
```diff
@@ -354,6 +357,9 @@ To create a dedicated {transform} node in the {default-dist}, set:
 ----
 
+If you want the node to use {ccs}, add `remote_cluster_client` to the list of
+roles. See <<remote-node>>.
```
```suggestion
roles. See <<remote-node>>.  A node with both the `transform` and `remote_cluster_client` roles
but no others is still considered a dedicated {transform} node.
```

droberts195 commented on docs/reference/modules/node.asciidoc:
```diff
@@ -340,6 +340,9 @@ xpack.ml.enabled: true <1>
 
+If you want the node to use {ccs}, add `remote_cluster_client` to the list of
+roles. See <<remote-node>>.
+
```
```suggestion

If you are using {ccs} in any way it is strongly recommended to allow {ml} nodes to do remote searches unless you have a very good reason not to. It will likely lead to confusion and frustration if cross cluster searches usually work perfectly but fail when used in {ml} job or {dfeed} configs.
```

droberts195 commented on docs/reference/modules/node.asciidoc:
```diff
@@ -354,6 +357,9 @@ To create a dedicated {transform} node in the {default-dist}, set:
 
+If you want the node to use {ccs}, add `remote_cluster_client` to the list of
+roles. See <<remote-node>>.
+
```
```suggestion

If you are using {ccs} in any way it is strongly recommended to allow {transform} nodes to do remote searches unless you have a very good reason not to. It will likely lead to confusion and frustration if cross cluster searches usually work perfectly but fail when used in {transform} configs.
```

droberts195 reviewed (commented):
I think we should spell out a couple of things more explicitly:

1. Adding `remote_cluster_client` to a dedicated ML or transform node still keeps it "dedicated"
2. We strongly recommend to do this unless you have a _very_ good reason not to, because we have seen through several support cases that it's completely baffling to end users when their searches work perfectly from dev console but fail when used in ML jobs

You probably don't want to accept my suggestions verbatim as I don't think they're worded very well, but hopefully you can come up with some words that say the same but sound better.

lcawl committed:
[DOCS] Add recommendation to ML and transform node settings

lcawl commented:
Thanks for the feedback @droberts195 !  I've updated the node page, as well as the ML settings and transform settings pages. I also carried over a clarification that data nodes are by default transform nodes, since that seemed to be missing from the transform settings page.  If that detail is no longer correct, however, I can remove it from both locations.

droberts195 commented on docs/reference/settings/transform-settings.asciidoc:
```diff
@@ -17,13 +17,18 @@ default.
 (<<static-cluster-setting,Static>>) Set `node.roles` to contain `transform` to
 identify the node as a _transform node_. If you use the default settings for a
-node, it does not have the `transform` role.
+node, it does not have the `transform` role. However, by default all generic
+data nodes are also {transform} nodes.
```
This is now very complicated.  I don't think you should use the phrase "generic data nodes", because what does that mean?

If you specify `node.data: true`, don't specify `node.roles` at all and don't specify `node.transform` at all then the node is a transform node.  (`node.data` and `node.transform` are old deprecated settings that will not work at all in 8.x.  So the statement used to make sense before `node.roles` was introduced.)

If you specify `node.roles: [ data ]` then the node is _not_ a transform node.

If you don't specify any node role settings (old or new) then the node is both a data node and a transform node (and every other type of node).

A data/transform node that specifies `node.roles` needs to explicitly say both, for example, `node.roles: [ data, transform ]` or `node.roles: [ data_hot, transform, remote_cluster_client ]`.

lcawl replied:
Thanks for the explanation!  I've removed "All data nodes are also transform nodes" from the node settings page and re-worked the details in the transform and ML settings pages to try to make this clearer.

lcawl committed:
[DOCS] Clarifies default transform node behaviour

droberts195 commented on docs/reference/modules/node.asciidoc:
```diff
@@ -15,7 +15,7 @@ All nodes know about all the other nodes in the cluster and can forward client
 
 By default, a node is all of the following types: master-eligible, data, ingest,
-and (if available) machine learning. All data nodes are also transform nodes.
+{transform} and (if available) {ml}.
```
```suggestion
{transform}, remote_cluster_client and (if available) {ml}.
```

droberts195 reviewed (approved):
The `remote_cluster_client` role is also on by default if you don't have any node-related settings in your config.

But LGTM apart from that so feel free to merge without further review.

robin13 commented on docs/reference/settings/ml-settings.asciidoc:
```diff
@@ -22,15 +22,19 @@ hardware, you must disable {ml} (by setting `xpack.ml.enabled` to `false`).
-If you use the `node.roles` setting, then all required roles must be explicitly
-set. Consult <<modules-node>> to learn more.
+By default, every node is a {ml} node. If you set `node.roles`, however,
+you must explicitly specify all the required roles for the node. To learn more, 
```
I think this sentence:

> If you set `node.roles`, however, you must explicitly specify all the required roles for the node.

Would be better in `docs/reference/modules/node.asciidoc`

lcawl replied:
I think it's already covered in the following sentence on that page: "If you set `node.roles`, the node is assigned only the roles you specify."

lcawl committed:
Update docs/reference/modules/node.asciidoc

Co-authored-by: David Roberts <dave.roberts@elastic.co>

lcawl committed:
[DOCS] Removes duplicate list of default node roles

lcawl commented:
> The `remote_cluster_client` role is also on by default if you don't have any node-related settings in your config.

On closer examination, the list of default roles is duplicated in this section: https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-node.html#node-roles, so I've added the transform role there and removed the earlier (redundant) paragraph.

lcawl committed:
[DOCS] Fixes typo

elasticmachine commented:
Pinging @elastic/es-docs (Team:Docs)

sophiec20 commented:
> `By default, every node is a {ml} node.`

Can we please drop this sentence (and its transform twin)?

I believe this adds confusion. In reality it means `by default every node is an ML node providing you have not set node.roles to anything, even an empty array.`. It gives too much prominence to the default settings of the dev env in the past. 

Imagine the user who has orchestration in place to deploy their cluster. It is likely, going forwards, that this orchestration would deploy with `node.roles` set. For this person, when they come to read this in the docs, in their frame of reference their default includes `node.roles`. Therefore we should tell users how to set `node.roles` in the most effective way for the production use case. We should not lead with the legacy or dev behaviour.
